srml/contracts/src/wasm/runtime.rs

```diff
@@ -344,12 +393,10 @@ define_env!(Env, <E: Ext>,
 			read_sandbox_memory_as(ctx, callee_ptr, callee_len)?;
 		let value: BalanceOf<<E as Ext>::T> =
 			read_sandbox_memory_as(ctx, value_ptr, value_len)?;
-		let input_data = read_sandbox_memory(ctx, input_data_ptr, input_data_len)?;
 
-		// Grab the scratch buffer and put in its' place an empty one.
-		// We will use it for creating `EmptyOutputBuf` container for the call.
-		let scratch_buf = mem::replace(&mut ctx.scratch_buf, Vec::new());
-		let empty_output_buf = EmptyOutputBuf::from_spare_vec(scratch_buf);
+		// Read input data into the scratch buffer, then take ownership of it.
+		read_sandbox_memory_into_scratch(ctx, input_data_ptr, input_data_len)?;
+		let input_data = mem::replace(&mut ctx.scratch_buf, Vec::new());
```

Comment:
the intereset of using scratch buffer here is to clean it at the same time ? why not read_sandbox_memory to input_data directly ?

Reply:
The purpose is to reduce memory allocations. `read_sandbox_memory` creates a new `Vec`, but the `scratch_buffer` is already available and its contents will be thrown out anyway.

Reply:
yes make sense